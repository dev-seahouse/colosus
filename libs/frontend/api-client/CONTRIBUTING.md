# Contributing

## Getting started

Before contributing, do make sure that you have fulfilled [pre-requisites](../../../README.md) for the project.

## Folder Structure

```
src/
  |- libs/
  | |- ApiLibraryApi
  | | |- <API library API>
  | |- stores/
  | | |- <zustand-based API store>
  | |- utils/
  | | |- <global level API utilities>
```

## Development

### API Reference

The API reference is generated using [Swagger](https://swagger.io/) and maintained by the Backend team.

In order to start the API reference locally, you need to serve colossus app via nx console or run `yarn nx run colossus:serve`
command from the root of the project.

Once the colossus server is running, you can access the API reference at [http://localhost:9000/openapi](http://localhost:9000/openapi).

### Shared Types

The shared types are maintained by the Backend team and are available as part of [@bambu/shared](../../shared/README.md) package.

Normally, you can import the shared types like this:

```typescript
import type { ISharedType } from '@bambu/shared';
```
