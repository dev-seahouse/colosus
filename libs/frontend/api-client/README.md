[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=frontend-api-client-old&metric=ncloc&token=73750ac968cd739da2de8b4d7b20df35786395d4)](https://sonarcloud.io/summary/new_code?id=frontend-api-client-old)
[![Bugs](https://sonarcloud.io/api/project_badges/measure?project=frontend-api-client-old&metric=bugs&token=73750ac968cd739da2de8b4d7b20df35786395d4)](https://sonarcloud.io/summary/new_code?id=frontend-api-client-old)
[![Vulnerabilities](https://sonarcloud.io/api/project_badges/measure?project=frontend-api-client-old&metric=vulnerabilities&token=73750ac968cd739da2de8b4d7b20df35786395d4)](https://sonarcloud.io/summary/new_code?id=frontend-api-client-old)
[![Code Smells](https://sonarcloud.io/api/project_badges/measure?project=frontend-api-client-old&metric=code_smells&token=73750ac968cd739da2de8b4d7b20df35786395d4)](https://sonarcloud.io/summary/new_code?id=frontend-api-client-old)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=frontend-api-client-old&metric=security_rating&token=73750ac968cd739da2de8b4d7b20df35786395d4)](https://sonarcloud.io/summary/new_code?id=frontend-api-client-old)
[![Maintainability Rating](https://sonarcloud.io/api/project_badges/measure?project=frontend-api-client-old&metric=sqale_rating&token=73750ac968cd739da2de8b4d7b20df35786395d4)](https://sonarcloud.io/summary/new_code?id=frontend-api-client-old)

# BAMBU API Client

This library contains the collection of APIs and API utilities usually required to start a BAMBU project.

## Installation

```
TBD
```

## Usage

### Setting Up Your API Store

```
// main.tsx
import * as ReactDOM from 'react-dom/client';
import { setupApiStore } from '@bambu/lib/frontend-api-client-old';

setupApiStore({ baseURL: 'BASE_API_URL' }).then(() => {
  // start your app
  const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement);

  root.render(<App />);
});
```

By setting up your store this way, any axios instance initialised with `createBambuAxiosInstance` function will automatically use the `baseURL` provided.

### Creating Custom API Instance

```typescript
// api.ts
import { createBambuAxiosInstance } from '@bambu/lib/frontend-api-client-old';

const api = createBambuAxiosInstance();

// this will automatically use the baseURL set in api store
const getCustomers = async () => api.get('/customers');
```

### Using The API

#### API Library API

##### Country Metadata API

Initialising the API instance:

```typescript
import { CountryMetadataApi } from '@bambu/frontend-api-client-old';

const countryApi = new CountryMetadataApi();
```

List of available Country Metadata API methods:

| **API**        | Method | **Description**                                                                                   |
| -------------- | ------ | ------------------------------------------------------------------------------------------------- |
| getCountries() | _/GET_ | To view a list of all the countries available in the platform and the dialing code, currency code |

#### Connect Advisor API

##### Auth API

Initialising the API instance:

```typescript
import { ConnectAdvisorAuthApi } from '@bambu/frontend-api-client-old';

const authApi = new ConnectAdvisorAuthApi();
```

List of available Country Metadata API methods:

| **API** | Method  | **Description**                                                                                                                          |
| ------- | ------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| login() | _/POST_ | Logs an advisor in on the Connect platform. It should perform the same action as the core /login endpoint, but with an inferred realmId. |

## Contributing

Please do read [Contributing](./CONTRIBUTING.md) for details.
