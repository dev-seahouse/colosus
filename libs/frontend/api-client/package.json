{
  "name": "@bambu/api-client",
  "version": "0.0.1",
  "type": "commonjs"
}
