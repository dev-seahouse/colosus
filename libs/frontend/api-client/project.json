{
  "name": "api-client",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/frontend/api-client/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/vite:build",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/frontend/api-client"
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs api-client {args.ver} {args.tag}"
      },
      "dependsOn": ["build"]
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{workspaceRoot}/coverage/libs/frontend/api-client"],
      "options": {
        "passWithNoTests": true,
        "reportsDirectory": "../../../coverage/libs/frontend/api-client"
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/frontend/api-client/**/*.ts"]
      }
    }
  },
  "tags": []
}
