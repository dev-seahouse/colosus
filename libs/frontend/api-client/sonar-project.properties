sonar.projectKey=api-client
sonar.organization=bambu-developer
sonar.sources=src/
sonar.exclusions=**/*.spec.*,mocks,**/handlers.ts,**/constants.ts,**/*.constants.ts
