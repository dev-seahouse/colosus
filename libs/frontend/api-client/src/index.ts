export * from './lib/store/useApiStore';
export * from './lib/store/useApiStore.selectors';
export * from './lib/utils/createBambuAxiosInstance/createBambuAxiosInstance';
export * from './lib/utils/setupApiStoreWithOriginOverride/setupApiStoreWithOriginOverride';

export * from './lib/Authentication/Authentication';

export * from './lib/BambuApiLibraryIntegration/CountryMetadata/CountryMetadata';
export * from './lib/BambuApiLibraryIntegration/Education/Education';
export * from './lib/BambuApiLibraryIntegration/Graph/Graph';
export * from './lib/BambuApiLibraryIntegration/House/House';
export * from './lib/BambuApiLibraryIntegration/Retirement/Retirement';

export * from './lib/ConnectInvestor/GoalTypes/GoalTypes';
export * from './lib/ConnectInvestor/AdvisorProfile/AdvisorProfile';
export * from './lib/ConnectInvestor/Leads/Leads';
export * from './lib/ConnectInvestor/LegalDocument/LegalDocument';
export * from './lib/ConnectInvestor/ModelPortfolios/ModelPortfolios';
export * from './lib/ConnectInvestor/InvestorPortfolios/InvestorPortfolio';
export * from './lib/ConnectInvestor/RiskProfiling/RiskProfiling';
export * from './lib/ConnectInvestor/RiskProfiling/RiskProfiling.schema';

export * from './lib/ConnectAdvisor/Auth/Auth';
export * from './lib/ConnectAdvisor/BrandAndSubdomain/BrandAndSubdomain';
export * from './lib/ConnectAdvisor/Branding/Branding';
export * from './lib/ConnectAdvisor/ContactMe/ContactMe';
export * from './lib/ConnectAdvisor/GoalTypes/GoalTypes';
export * from './lib/ConnectAdvisor/LegalDocument/LegalDocument';
export * from './lib/ConnectAdvisor/PortfolioSummary/PortfolioSummary';
export * from './lib/ConnectAdvisor/Profile/Profile';
export * from './lib/ConnectAdvisor/ProfileBio/ProfileBio';
export * from './lib/ConnectAdvisor/TopLevelOptions/TopLevelOptions';
export * from './lib/ConnectAdvisor/Leads/Leads';
export * from './lib/ConnectAdvisor/Leads/mocks/mockResponse';
export * from './lib/ConnectAdvisor/Preferences/Preferences';
export * from './lib/ConnectAdvisor/RiskProfiling/RiskProfiling';

export * from './lib/StripeIntegration/AdvisorSubscriptions/AdvisorSubscriptions';
export * from './lib/StripeIntegration/BillingPortal/BillingPortal';
export * from './lib/StripeIntegration/Prices/Prices';

export * from './lib/Tenant/Branding/Branding';

export * from './lib/TransactInvestor/Auth/Auth';
export * from './lib/TransactInvestor/Authenticated/Profile/Profile';
export * from './lib/TransactInvestor/Authenticated/Profile/mocks/getInvestorPlatformUserProfileMockResponse';
export * from './lib/TransactInvestor/ModelPortfolio/ModelPortfolio';

export * from './lib/TransactAdvisor/Instruments/Instruments';
export * from './lib/TransactAdvisor/ModelPortfolio/ModelPortfolio';
export * from './lib/TransactAdvisor/AdvisorBankAccount/AdvisorBankAccount';
export * from './lib/TransactAdvisor/AdvisorBankAccount/AdvisorBankAccount.utils';

export * from './lib/TransactInvestor/Authenticated/Brokerage/Brokerage';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/kycBrokerageMockResponse';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/submitKycToBrokerageMockRequest';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/getDirectDebitMandateMockRes';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/createBankAccMockReq';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/createBankAccountMockRes';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/createDirectDebitMandateMockReq';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/createDirectDebitMandateMockRes';
export * from './lib/TransactInvestor/Authenticated/Brokerage/mocks/createDirectDebitSubscriptionMockReq';
export * from './lib/TransactInvestor/Authenticated/Brokerage/BankAccounts/BankAccountsV2';

export * from './lib/TransactInvestor/Instruments/Instruments';
export * from './lib/TransactInvestor/Instruments/mocks/getInstrumentsMockResponse';
export * from './lib/TransactInvestor/Instruments/mocks/getInstrumentAssetClassesMockResponse';

export * from './mocks/handlers';
export * from './mocks/getMockBaseApiUrl';
export * from './lib/TransactAdvisor/Brokerages/Brokerages';
export * from './lib/TransactInvestor/Authenticated/Brokerage/Brokerage.types';
export * from './lib/TransactInvestor/Authenticated/Brokerage/Brokerage.enum';
export * from './lib/TransactInvestor/Authenticated/Brokerage/Brokerage.utils';
export * from './lib/types';

export * from './lib/TransactInvestor/Authenticated/Goals/Goals';
