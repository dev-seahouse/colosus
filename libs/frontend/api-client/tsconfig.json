{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "target": "esnext",
    "moduleResolution": "node",
    "module": "esnext",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "types": ["vitest", "vite/client"]
  },
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ]
}
