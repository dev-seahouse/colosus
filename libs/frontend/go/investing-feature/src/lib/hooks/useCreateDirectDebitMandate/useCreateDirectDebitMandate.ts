import type {
  InvestorBrokerageCreateDirectDebitMandateRequestDto,
  InvestorBrokerageCreateDirectDebitMandateResponseDto,
} from '@bambu/api-client';
import { TransactInvestorAuthenticatedBrokerageApi } from '@bambu/api-client';
import { useMutation, type UseMutationOptions } from '@tanstack/react-query';
import type { AxiosError } from 'axios';

export function useCreateDirectDebitMandate(
  queryOptions?: UseMutationOptions<
    InvestorBrokerageCreateDirectDebitMandateResponseDto,
    AxiosError,
    InvestorBrokerageCreateDirectDebitMandateRequestDto
  >
) {
  return useMutation({
    mutationKey: ['createDirectDebitMandate'],
    mutationFn: (args: InvestorBrokerageCreateDirectDebitMandateRequestDto) =>
      postToDirectDebitMandate(args),
    ...queryOptions,
  });
}

async function postToDirectDebitMandate(
  args: InvestorBrokerageCreateDirectDebitMandateRequestDto
) {
  const createDiDirectDebitMandateApi =
    new TransactInvestorAuthenticatedBrokerageApi();
  const res =
    await createDiDirectDebitMandateApi.createDirectDebitMandateForParty(args);
  return res.data;
}

export default useCreateDirectDebitMandate;
