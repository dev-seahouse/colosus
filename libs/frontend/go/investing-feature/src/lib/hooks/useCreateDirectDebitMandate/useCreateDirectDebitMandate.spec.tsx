import useCreateDirectDebitMandate from './useCreateDirectDebitMandate';
import {
  act,
  queryClientWrapper,
  renderHook,
  waitFor,
} from '@bambu/react-test-utils';
import { createDirectDebitMandateMockReq } from '@bambu/api-client';

describe('useCreateDirectDebitMandate', () => {
  const mockOnSuccess = vi.fn();

  afterAll(() => {
    vi.resetAllMocks();
  });

  it('should render successfully', async () => {
    const { result } = renderHook(
      () =>
        useCreateDirectDebitMandate({
          onSuccess: mockOnSuccess,
        }),
      {
        wrapper: queryClientWrapper,
      }
    );
    act(() => result.current.mutate(createDirectDebitMandateMockReq));
    await waitFor(() => result.current.isSuccess);
    expect(mockOnSuccess).toHaveBeenCalled();
    expect(result.current.data).toMatchSnapshot();
  });
});
