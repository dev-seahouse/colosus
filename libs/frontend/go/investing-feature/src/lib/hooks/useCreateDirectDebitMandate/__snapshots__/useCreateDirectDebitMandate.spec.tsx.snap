// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useCreateDirectDebitMandate > should render successfully 1`] = `
{
  "bankAccountId": "string",
  "id": "string",
  "partyId": "string",
  "reason": "string",
  "reference": "string",
  "status": "Pending",
}
`;
